src/kOS.Safe/Encapsulation/ScalarValue.cs

```diff
@@ -0,0 +1,456 @@
+                {
+                    double d = GetDoubleValue();
+                    if (double.IsInfinity(d) || double.IsNaN(d)) return false;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public object Value { get { return internalValue; } }
+
+        protected ScalarValue()
+        {
+            InitializeSuffixes();
+        }
+
+        public void InitializeSuffixes()
+        {
+            AddSuffix("ISINT", new Suffix<bool>(() => IsInt));
+            AddSuffix("ISDOUBLE", new Suffix<bool>(() => IsDouble));
+            AddSuffix("ISVALID", new Suffix<bool>(() => IsValid));
+        }
+
+        public static ScalarValue Create(object value)
+        {
+            if (value is float)
+                value = Convert.ToDouble(value);
+            if (value is double)
+            {
+                bool inBounds = Int32.MinValue < (double)value && (double)value < Int32.MaxValue;
+                if (inBounds && !double.IsNaN((double)value))
+                {
+                    // Convert the double to an int, and check and see if they are still equal.
+                    // if so, treat the double as if it was an int.
+                    int intPart = Convert.ToInt32(value);
+                    if ((double)value == intPart)
+                    {
+                        return new ScalarIntValue(intPart);
+                    }
+                }
+                return new ScalarDoubleValue((double)value);
+            }
+            else if (value is int)
+            {
+                return new ScalarIntValue((int)value);
+            }
+            else if (value is ScalarValue)
+            {
+                return ScalarValue.Create(((ScalarValue)value).Value);
+            }
+            else
+            {
+                throw new kOS.Safe.Exceptions.KOSException(string.Format("Failed to set scalar value.  Passed type {0}, expected Double or Int", value.GetType().Name));
+            }
+        }
+
+        public int GetIntValue()
+        {
+            return Convert.ToInt32(Value);
+        }
+
+        public double GetDoubleValue()
+        {
+            return Convert.ToDouble(Value);
+        }
+
+        public override string ToString()
+        {
+            if (IsInt) return GetIntValue().ToString();
+            else if (IsDouble) return GetDoubleValue().ToString();
+            return "NaN";
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null) return false;
+            if (obj is ScalarValue)
+            {
+                ScalarValue val = obj as ScalarValue;
+                if (val != null)
+                {
+                    if (this.IsInt && val.IsInt)
+                    {
+                        return this.GetIntValue() == val.GetIntValue();
+                    }
+                    return this.GetDoubleValue() == val.GetDoubleValue();
```

Comment:
You two understand floating point math itself better than I.  If there is no loss of precision, I would suggest we just treat everything like an double.  There is nowhere in the code that we limit the operations based on being an integer, and in fact operations like `Power` and `Divide` require us to treat all values like doubles already.  `Divide` is probably the most notable, since the value is very different in integer math than it is in floating point math (`3/2 = 1`, rather than `1.5`).  But it would be important that the sum of the integer values would equal an integer parse (i.e. `100000 + 228000 = 328000` evaluates to `true`).  Since I'm not familiar with the inner workings of floating points, I don't know at which point that becomes an issue, maybe it only breaks down beyond the storage capacity of the integer anyways.

I should also point out that right now there is nothing to address overflow, so I think it is possible to add two positive integer values and return a negative integer value.